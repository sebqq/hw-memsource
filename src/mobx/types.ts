export type FetchState = "stale" | "pending" | "error" | "expired";
