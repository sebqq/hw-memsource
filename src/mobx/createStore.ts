import { IAnyModelType, Instance, onSnapshot, types } from "mobx-state-tree";

import { AuthStore, AUTH_INIT_STATE } from "./entities/auth/AuthStore";
import {
  ProjectStore,
  PROJECTS_INIT_STATE,
} from "./entities/projects/ProjectStore";

export type RootStoreModel = Instance<typeof RootStore>;

export const RootStore = types
  .model("Store", {
    authStore: types.late((): IAnyModelType => AuthStore),
    projectStore: types.late((): IAnyModelType => ProjectStore),
  })
  .views((self) => ({
    get accessToken(): string | null {
      return self.authStore.token;
    },
  }))
  .actions((self) => {
    return {
      resetAll() {
        self.authStore.reset("stale");
        self.projectStore.reset("stale");
      },
    };
  });

export function createStore() {
  const store = RootStore.create({
    authStore: AUTH_INIT_STATE,
    projectStore: PROJECTS_INIT_STATE,
  });

  // Listen to new snapshots, which are created anytime something changes
  onSnapshot(store.authStore, (snapshot) => {
    store.authStore.persist(snapshot);
  });

  return store;
}

export type TStore = ReturnType<typeof createStore>;
