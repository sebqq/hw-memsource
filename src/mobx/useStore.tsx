import React from "react";
import { useLocalObservable } from "mobx-react-lite";

import { createStore, RootStoreModel } from "./createStore";
import { AuthStoreModel } from "./entities/auth/AuthStore";
import { ProjectStoreModel } from "./entities/projects/ProjectStore";

export interface StoreContextType extends RootStoreModel {
  authStore: AuthStoreModel;
  projectStore: ProjectStoreModel;
}

const storeContext = React.createContext<StoreContextType | null>(null);

export const StoreProvider: React.FC = ({ children }) => {
  const store = useLocalObservable(createStore);
  return (
    <storeContext.Provider value={store}>{children}</storeContext.Provider>
  );
};

export const useStore = () => {
  const store = React.useContext(storeContext);
  if (!store) {
    // this is especially useful in TypeScript so you don't need to be checking for null all the time
    throw new Error("useStore must be used within a StoreProvider.");
  }
  return store;
};
