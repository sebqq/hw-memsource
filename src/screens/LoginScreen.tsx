import { observer } from "mobx-react-lite";
import React, { useRef, useState } from "react";
import { Text, View, StyleSheet, TextInput as RNTextInput } from "react-native";
import * as WebBrowser from "expo-web-browser";

import TextInput from "../components/Form/TextInput";
import ContainerLoadingIndicator from "../components/Loading/ContainerLoadingIndicator";
import { useStore } from "../mobx/useStore";
import theme from "../utils/theme";
import Button from "../components/Button";
import ErrorText from "../components/Form/ErrorText";
import { ScrollView } from "react-native-gesture-handler";

const LoginScreen = () => {
  const { authStore } = useStore();

  const usernameInputRef = useRef<RNTextInput>(null);
  const passwordInputRef = useRef<RNTextInput>(null);

  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");

  async function onLoginPressed() {
    if (username === "") {
      usernameInputRef.current?.focus();
      return;
    }
    if (password === "") {
      passwordInputRef.current?.focus();
      return;
    }
    await authStore.signIn(username, password);
  }

  function onResetPasswordPressed() {
    WebBrowser.openBrowserAsync(
      "https://cloud.memsource.com/web/passwordRecovery/send"
    );
  }

  return (
    <>
      <ScrollView testID="Login.container" style={styles.innerContainer}>
        <Text style={styles.headingLabel}>
          You will need to{" "}
          <Text style={theme.styles.fontBold}>sign in first</Text> in order to
          see and/or manage your projects.
        </Text>
        <TextInput
          testID="Login.username"
          inputRef={usernameInputRef}
          labelText="Your Username:"
          placeholder="Peter"
          textContentType="nickname"
          value={username}
          onChangeText={setUsername}
        />
        <TextInput
          testID="Login.password"
          inputRef={passwordInputRef}
          labelText="Your Password:"
          placeholder="*********"
          textContentType="password"
          value={password}
          onChangeText={setPassword}
          secureTextEntry
        />
        <ErrorText
          testID="Login.Error"
          error={authStore.requestState.state === "error"}
          errorText={authStore.requestState.errorMessage}
        />
        <View style={styles.spacer} />
        <Button
          testID="Login.submit"
          text="Sign in"
          onPress={onLoginPressed}
          disabled={authStore.requestState.state === "pending"}
        />
        <View style={styles.spacer} />
        <Text style={styles.secondaryButtonLabel}>Forgot your password?</Text>
        <Text
          onPress={onResetPasswordPressed}
          style={styles.secondaryButtonLink}
        >
          Reset Password
        </Text>
      </ScrollView>
      {authStore.requestState.state === "pending" && (
        <ContainerLoadingIndicator text="Logging in..." />
      )}
    </>
  );
};

const styles = StyleSheet.create({
  innerContainer: {
    marginHorizontal: 16,
    paddingTop: 16,
    position: "relative",
  },
  spacer: {
    marginTop: 20,
  },
  headingLabel: {
    fontSize: 22,
    color: theme.colors.gray[700],
    marginBottom: 12,
  },
  errorText: {
    marginTop: 10,
    color: theme.colors.red[400],
    fontWeight: "bold",
  },
  secondaryButtonLabel: {
    fontSize: 16,
    marginTop: 6,
    textAlign: "center",
    color: theme.colors.gray[500],
  },
  secondaryButtonLink: {
    marginTop: 4,
    fontSize: 16,
    textAlign: "center",
    color: theme.colors.blue[600],
    textDecorationLine: "underline",
  },
});

export default observer(LoginScreen);
