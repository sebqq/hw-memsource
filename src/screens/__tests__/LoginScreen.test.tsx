import * as React from "react";
import { render, fireEvent, act } from "@testing-library/react-native";
import fetchMock from "jest-fetch-mock";
import { when } from "mobx";
import { Instance } from "mobx-state-tree";

import { useStore as storeHook } from "../../mobx/useStore";
import {
  AuthStore,
  AuthStoreModel,
  AUTH_INIT_STATE,
} from "../../mobx/entities/auth/AuthStore";
import LoginScreen from "../LoginScreen";

const useStore = storeHook as ReturnType<typeof jest["fn"]>;
jest.mock("../../mobx/useStore");

let authStore: Instance<AuthStoreModel>;

const stubAuthStore = () => {
  authStore = AuthStore.create(AUTH_INIT_STATE);
  return { authStore };
};

describe("LoginScreen tests", () => {
  beforeEach(() => {
    global.fetch = fetchMock;
    useStore.mockReturnValue(stubAuthStore());
    fetchMock.resetMocks();
  });

  it("it renders all inputs as expected", () => {
    const { toJSON } = render(<LoginScreen />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("Login - error flow", async () => {
    fetchMock.mockResponseOnce(JSON.stringify(""), { status: 401 });

    const { getByText, getByTestId } = render(<LoginScreen />);
    expect(getByTestId("Login.Error").props.children).toBeNull();

    fireEvent(getByTestId("Login.username"), "onChangeText", "testUsername");
    fireEvent(getByTestId("Login.password"), "onChangeText", "testPasswd");

    fireEvent.press(getByText("Sign in"));
    await act(async () => {
      await when(() => authStore.requestState.state === "error");
    });
    expect(getByTestId("Login.Error").props.children).not.toBeNull();
  });

  it("Login - success flow", async () => {
    fetchMock.mockResponseOnce(
      JSON.stringify({
        token: "123456",
        expires: "2021-01-150T04:00:00-0000",
        user: {
          firstName: "firstName",
          lastName: "lastname",
          userName: "username",
          email: "email@email.com",
          role: "NONE",
          id: "1",
          uid: "1",
        },
      }),
      { status: 200 }
    );

    const { getByText, getByTestId } = render(<LoginScreen />);
    expect(getByTestId("Login.Error").props.children).toBeNull();

    fireEvent(getByTestId("Login.username"), "onChangeText", "testUsername");
    fireEvent(getByTestId("Login.password"), "onChangeText", "testPasswd");

    fireEvent.press(getByText("Sign in"));
    await act(async () => {
      await when(() => authStore.requestState.state === "stale");
    });
    expect(getByTestId("Login.Error").props.children).toBeNull();
  });
});
