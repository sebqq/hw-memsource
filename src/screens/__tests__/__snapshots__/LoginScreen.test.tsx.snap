// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen tests it renders all inputs as expected 1`] = `
<RCTScrollView
  collapsable={false}
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  style={
    Object {
      "marginHorizontal": 16,
      "paddingTop": 16,
      "position": "relative",
    }
  }
  testID="Login.container"
>
  <View>
    <Text
      style={
        Object {
          "color": "#2D3748",
          "fontSize": 22,
          "marginBottom": 12,
        }
      }
    >
      You will need to
       
      <Text
        style={
          Object {
            "fontWeight": "600",
          }
        }
      >
        sign in first
      </Text>
       in order to see and/or manage your projects.
    </Text>
    <View
      style={
        Array [
          Object {
            "marginVertical": 10,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#4A5568",
              "fontSize": 12,
              "fontWeight": "bold",
              "letterSpacing": 1,
              "marginBottom": 10,
              "textTransform": "uppercase",
            },
            undefined,
          ]
        }
      >
        Your Username:
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="Peter"
        placeholderTextColor="#A0AEC0"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "backgroundColor": "white",
              "borderColor": "white",
              "borderRadius": 7,
              "borderWidth": 2,
              "color": "#2D3748",
              "fontSize": 16,
              "height": 57,
              "paddingHorizontal": 10,
            },
            undefined,
            undefined,
          ]
        }
        testID="Login.username"
        textContentType="nickname"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "marginVertical": 10,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#4A5568",
              "fontSize": 12,
              "fontWeight": "bold",
              "letterSpacing": 1,
              "marginBottom": 10,
              "textTransform": "uppercase",
            },
            undefined,
          ]
        }
      >
        Your Password:
      </Text>
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        placeholder="*********"
        placeholderTextColor="#A0AEC0"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "backgroundColor": "white",
              "borderColor": "white",
              "borderRadius": 7,
              "borderWidth": 2,
              "color": "#2D3748",
              "fontSize": 16,
              "height": 57,
              "paddingHorizontal": 10,
            },
            undefined,
            undefined,
          ]
        }
        testID="Login.password"
        textContentType="password"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <Text
      style={
        Object {
          "color": "#F56565",
          "fontWeight": "bold",
        }
      }
      testID="Login.Error"
    />
    <View
      style={
        Object {
          "marginTop": 20,
        }
      }
    />
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#3182ce",
          "borderRadius": 8,
          "opacity": 1,
          "paddingHorizontal": 45,
          "paddingVertical": 12,
        }
      }
      testID="Login.submit"
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 18,
            "fontWeight": "500",
          }
        }
      >
        Sign in
      </Text>
    </View>
    <View
      style={
        Object {
          "marginTop": 20,
        }
      }
    />
    <Text
      style={
        Object {
          "color": "#718096",
          "fontSize": 16,
          "marginTop": 6,
          "textAlign": "center",
        }
      }
    >
      Forgot your password?
    </Text>
    <Text
      onPress={[Function]}
      style={
        Object {
          "color": "#2b6cb0",
          "fontSize": 16,
          "marginTop": 4,
          "textAlign": "center",
          "textDecorationLine": "underline",
        }
      }
    >
      Reset Password
    </Text>
  </View>
</RCTScrollView>
`;
